hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/router/RouterClientProtocol.java

```diff
@@ -1928,9 +1929,17 @@ public BatchedEntries<OpenFileEntry> listOpenFiles(long prevId,
   @Override
   public void msync() throws IOException {
     rpcServer.checkOperation(NameNode.OperationCategory.READ, true);
-    Set<FederationNamespaceInfo> nss = namenodeResolver.getNamespaces();
+    Set<FederationNamespaceInfo> allNamespaces = namenodeResolver.getNamespaces();
     RemoteMethod method = new RemoteMethod("msync");
-    rpcClient.invokeConcurrent(nss, method);
+    Set<FederationNamespaceInfo> namespacesEligibleForObserverReads = new HashSet<>();
```

Comment:
How often does this change? Seems like the namenodeResolver could do this calculation once and cache.

Reply:
allNamespace does not change often. Only when an entire nameservice becomes inactive. The set of nonEligibleNamenodes is only refreshed on startup. A cache may help. Wondering whether checking the cache is just as expensive as doing the calculation each time. A set equality check vs. filtering the set each time.

Reply:
Added a cache.